sanitize.go

```diff
@@ -0,0 +1,143 @@
+package dns
+
+import "strings"
+
+// Dedup removes identical RRs from rrs. It preserves the original ordering.
+// The lowest TTL of any duplicates is used in the remaining one.
+//
+// TODO(miek): This function will be extended to also look for CNAMEs and DNAMEs.
+// if found, it will prune rrs from the "other data" that can exist. Example:
+// if it finds a: a.miek.nl. CNAME foo, all other RRs with the ownername a.miek.nl.
+// will be removed. When a DNAME is found all RRs with an ownername below that of
+// the DNAME will be removed.
+func Dedup(rrs []RR) []RR {
+	m := make(map[string]RR)
+	keys := make([]string, 0, len(rrs))
+
+	// Save possible cname and dname domainnames. Currently a slice, don't
+	// expect millions here..
+	cname := []string{}
+	dname := []string{}
+
+	for _, r := range rrs {
+		key, end := normalizedString(r)
+		keys = append(keys, key)
+		if _, ok := m[key]; ok {
+			// Shortest TTL wins.
+			if m[key].Header().Ttl > r.Header().Ttl {
+				m[key].Header().Ttl = r.Header().Ttl
+			}
+			continue
+		}
+
+		if r.Header().Rrtype == TypeCNAME {
+			// we do end+3 here, so we capture the full domain name *and*
+			// the class field which mnemonic is always two chars.
+			cname = append(cname, key[:end+3])
```

Comment:
True. But it will still work, because there is no overlap. But indeed needs
commenting.
On 26 Aug 2015 13:13, "Alex Sergeyev" <notifications@github.com> wrote:

> In sanitize.go
> <https://github.com/miekg/dns/pull/255#discussion_r37974317>:
>
> > +
> > +	for _, r := range rrs {
> > +		key, end := normalizedString(r)
> > +		keys = append(keys, key)
> > +		if _, ok := m[key]; ok {
> > +			// Shortest TTL wins.
> > +			if m[key].Header().Ttl > r.Header().Ttl {
> > +				m[key].Header().Ttl = r.Header().Ttl
> > +			}
> > +			continue
> > +		}
> > +
> > +		if r.Header().Rrtype == TypeCNAME {
> > +			// we do end+3 here, so we capture the full domain name *and*
> > +			// the class field which mnemonic is always two chars.
> > +			cname = append(cname, key[:end+3])
>
> "always two letters" does not hold since there is ANY class :(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/miekg/dns/pull/255/files#r37974317>.
>